crates/bevy_math/src/curves/variable.rs

```diff
@@ -0,0 +1,230 @@
+//         }
+//     }
+
+//     // pub fn insert(&mut self, time_sample: f32, value: T) {
+//     // }
+
+//     // pub fn remove(&mut self, index: usize) {
+//     //assert!(samples.len() > 1, "curve can't be empty");
+//     // }
+
+//     pub fn add_offset_time(&mut self, time_offset: f32) {
+//         self.time_stamps.iter_mut().for_each(|t| *t += time_offset);
+//     }
+
+//     pub fn iter(&self) -> impl Iterator<Item = (f32, &T)> {
+//         self.time_stamps.iter().copied().zip(self.keyframes.iter())
+//     }
+
+//     pub fn iter_mut(&mut self) -> impl Iterator<Item = (f32, &mut T)> {
+//         self.time_stamps
+//             .iter()
+//             .copied()
+//             .zip(self.keyframes.iter_mut())
+//     }
+// }
+
+impl<T> Curve for CurveVariable<T>
+where
+    T: Interpolate + Clone + 'static,
+{
+    type Output = T;
+
+    fn duration(&self) -> f32 {
+        self.time_stamps.last().copied().unwrap_or(0.0)
+    }
+
+    fn sample(&self, time: f32) -> Self::Output {
+        // Index guessing gives a small search optimization
+        let index = if time < self.duration() * 0.5 {
+            0
+        } else {
+            self.time_stamps.len() - 1
+        };
+
+        self.sample_with_cursor(index as u16, time).1
+    }
+
+    fn sample_with_cursor(&self, mut cursor: u16, time: f32) -> (u16, Self::Output) {
+        // Adjust for the current keyframe cursor
+        let last_cursor = (self.time_stamps.len() - 1) as u16;
+
+        cursor = cursor.max(0).min(last_cursor);
+        if self.time_stamps[cursor as usize] < time {
+            // Forward search
+            loop {
+                if cursor == last_cursor {
+                    return (last_cursor, self.keyframes[last_cursor as usize].clone());
+                }
+                cursor += 1;
+
+                if self.time_stamps[cursor as usize] >= time {
+                    break;
+                }
+            }
+        } else {
+            // Backward search
+            loop {
+                if cursor == 0 {
+                    return (0, self.keyframes[0].clone());
+                }
+
+                let i = cursor - 1;
+                if self.time_stamps[i as usize] <= time {
+                    break;
+                }
+
+                cursor = i;
+            }
+        }
+
+        // Lerp the value
+        let i = cursor - 1;
+        let previous_time = self.time_stamps[i as usize];
+        let t = (time - previous_time) / (self.time_stamps[cursor as usize] - previous_time);
+        debug_assert!(
+            (0.0..=1.0).contains(&t),
+            "t = {} but should be normalized",
+            t
+        ); // Checks if it's required to normalize t
+
+        let value = T::interpolate(
+            &self.keyframes[i as usize],
+            &self.keyframes[cursor as usize],
+            &self.interpolations[i as usize],
+            t,
+        );
+
+        (cursor, value)
+    }
+}
+
+// #[cfg(test)]
```

Comment:
should this tests be commented?

Reply:
No actually I just redesigned the `CurveVariable` type, next commits I will be creating tests for it